### PR TITLE
Fix bug 1243086. Added direct links to the search, filtering and the translations.

### PR DESCRIPTION
**Pull request by jotes**
Unfortunately, I still need some work before we'll be able to review this

* [x] rebase with latest master
* [x] load application state from the url in an one query to the backend
* [x] modify backend view to load currently edited entity

### CODE REVIEW

**Comment by mathjazz** [comment]
Really nice work!

I've only found a few bugs, but they are all easy to fix:
- [ ] Rename querystring parameters `filterType` -> `filter`, `filterSearch` -> `search`.
- [ ] Filters should use the same names we use in the filter list (e.g. `suggested` instead of `translated`).
- [ ] Inplace: after filter is applied, entities are copied under the `Not on the current page list`.
- [ ] Inplace: if only filter and/or search params are specified, entity list gets updated, but sidebar remains closed. If entity params is specified, sidebar opens with editor editing the specified entity.

I'll have a closer look at the code later.

File: pontoon/base/static/js/translate.js [hunk 1]
```diff
@@ -1499,7 +1503,7 @@ var Pontoon = (function (my) {
      * Add entity translation to localStorage
      *
      * entity Entity
-     * translation Translation
+     * translation Translatio n
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: Remove space.

File: pontoon/base/static/js/translate.js [hunk 2]
```diff
@@ -1998,44 +2002,128 @@ var Pontoon = (function (my) {
...
+      }
+    },
+
+    /*
+     * Returns an entity object with given id.
+     */
+    getEntityById: function(entityId) {
+        return this.entities.find(function(entity){
+          return entity.pk == parseInt(entityId);
+        });
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: Remove double indent.

File: pontoon/base/static/js/translate.js [hunk 3]
```diff
@@ -2723,6 +2848,19 @@ var Pontoon = (function (my) {
...
+    getCurrentEntity: function() {
+      var $editor = $('#editor'),
+          $sidebar = $('#sidebar');
+
+      if ((this.requiresInplaceEditor() && !$editor.is('.opened')) ||
+          (!this.requiresInplaceEditor() && $sidebar.hasClass('no'))) {
+        return
+      }
+      return $editor[0].entity;
+    },
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: We should merge this with `Pontoon.getEditorEntity()`.

File: pontoon/base/static/js/translate.js [hunk 4]
```diff
@@ -2731,29 +2869,106 @@ var Pontoon = (function (my) {
...
+      };
+
+     if (forceUpdate || requestedPaths !== paths) {
+       self.updateCurrentUrl(this.state);
+     }
+    },
+
+    getSearchQuery: function() {
+      return $('#search').val();
+    },
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: We already have `Ponton.getSearch()`.

File: pontoon/base/static/js/translate.js [hunk 5]
```diff
@@ -2731,29 +2869,106 @@ var Pontoon = (function (my) {
...
+
+      if (!history.state || forceUpdate) {
+        this.state['filterType'] = queryFilterType;
+        this.state['filterSearch'] = queryFilterSearch;
+        this.state['currentEntity'] = queryCurrentEntity;
+      };
+
+     if (forceUpdate || requestedPaths !== paths) {
+       self.updateCurrentUrl(this.state);
+     }
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: Add indent.

**Comment by jotes** [comment]
@mathjazz could you check again?

File: pontoon/base/static/js/translate.js [hunk 6]
```diff
@@ -420,7 +420,10 @@ var Pontoon = (function (my) {
       if (!this.app.advanced) {
         $("#entitylist").css('left', -$('#sidebar').width());
         $("#editor").addClass('opened').css('left', 0);
+      } else {
+        $('#editor').addClass('opened');
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: Why do we need this?

File: pontoon/base/static/js/translate.js [hunk 7]
```diff
@@ -1998,44 +2002,128 @@ var Pontoon = (function (my) {
...
+      self.updateProgress();
       $("#progress").show();
 
       $("#project-load").hide();
 
-      // If 2-column layout opened by default, open first entity in the editor
-      if (this.app.advanced) {
-        this.openFirstEntity();
+      if (self.requiresInplaceEditor()) {
+        self.waitForAttribute('entities').then(function() {
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: I couldn't figure out why do we need to wait for `Pontoon.entities` here. At this stage, it should be always available.

File: pontoon/base/static/js/translate.js [hunk 8]
```diff
@@ -2169,15 +2257,17 @@ var Pontoon = (function (my) {
...
+          }, $.proxy(Pontoon.noEntitiesError, Pontoon));
           break;
 
         case "DATA":
           // Deep copy: http://api.jquery.com/jQuery.extend
-          Pontoon.entities = $.extend(
-            true,
-            Pontoon.entities,
-            message.value.entities);
+          Pontoon.waitForAttribute('entities').then(function() {
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: Or here.

File: pontoon/base/static/js/translate.js [hunk 9]
```diff
@@ -1998,44 +2002,128 @@ var Pontoon = (function (my) {
...
+        // If 2-column layout opened by default, open first entity in the editor
+        if (this.app.advanced) {
+          if (this.getEntitiesIds().length > 0){
+            $('#sidebar').removeClass('no');
+            this.openFirstEntity();
+          }
+          // If not and editor opened, show entity list
+        } else if ($("#editor").is('.opened')) {
+          this.goBackToEntityList();
+        }
```
Comment (mathjazz) [inline pontoon/base/static/js/translate.js]: Seems like the same block we use above.

**Comment by mathjazz** [comment]
Thanks @jotes, you've made a great progress here!

There are several new features I've discovered while testing and reviewing this, and some bugs. As we agreed on IRC, I'm closing this PR and creating a new one, based on it.

General advice: please keep double space between JS functions. :-) I know it's probably weird, but let's keep it consistent before we lint the entire code or something.

**Comment by jotes** [comment]
Okay!

File: pontoon/base/static/js/translate.js (hunk 6 above)
Comment (jotes) [inline pontoon/base/static/js/translate.js]: I think we can safely remove this line.

File: pontoon/base/static/js/translate.js (hunk 7 above)
Comment (jotes) [inline pontoon/base/static/js/translate.js]: I guess that some other problem caused that entities were empty at that point, so i decided to wait for them. I think we can remove this wait

File: pontoon/base/static/js/translate.js (hunk 8 above)
Comment (jotes) [inline pontoon/base/static/js/translate.js]: The same thing as above.

File: pontoon/base/static/js/translate.js (hunk 9 above)
Comment (jotes) [inline pontoon/base/static/js/translate.js]: i guess we definetely coud generalize that.

**Pull request closed without merging**